core/src/cubos/gl/debug_render.cpp

```diff
@@ -0,0 +1,471 @@
+
+            uniform vec3 objColor;
+
+            void main()
+            {
+                color = vec4(objColor, 1.0f);
+            }
+        )");
+
+    pipeline = renderDevice->createShaderPipeline(vs, ps);
+
+    initCube();
+    initSphere();
+
+    mvpBuffer = renderDevice->createConstantBuffer(sizeof(glm::mat4), nullptr, gl::Usage::Dynamic);
+    mvpBindingPoint = pipeline->getBindingPoint("MVP");
+    colorBindingPoint = pipeline->getBindingPoint("objColor");
+
+    RasterStateDesc rsDesc;
+
+    rsDesc.rasterMode = RasterMode::Fill;
+    fillRasterState = renderDevice->createRasterState(rsDesc);
+
+    rsDesc.rasterMode = RasterMode::Wireframe;
+    wireframeRasterState = renderDevice->createRasterState(rsDesc);
+}
+
+void Debug::drawCube(glm::vec3 center, glm::vec3 size, float time, glm::quat rotation, glm::vec3 color)
+{
+    debugDrawMutex.lock();
+    requests.push_back(DebugDrawRequest{
+        objCube, fillRasterState, glm::translate(center) * glm::toMat4(rotation) * glm::scale(size), time, color});
+    debugDrawMutex.unlock();
+}
+
+void Debug::drawWireCube(glm::vec3 center, glm::vec3 size, float time, glm::quat rotation, glm::vec3 color)
+{
+    debugDrawMutex.lock();
+    requests.push_back(DebugDrawRequest{
+        objCube, wireframeRasterState, glm::translate(center) * glm::toMat4(rotation) * glm::scale(size), time, color});
+    debugDrawMutex.unlock();
+}
+
+void Debug::drawSphere(glm::vec3 center, float radius, float time, glm::vec3 color)
+{
+    debugDrawMutex.lock();
+    requests.push_back(DebugDrawRequest{objSphere, fillRasterState,
+                                        glm::translate(center) * glm::scale(glm::vec3(radius)), time, color});
+    debugDrawMutex.unlock();
+}
+
+void Debug::drawWireSphere(glm::vec3 center, float radius, float time, glm::vec3 color)
+{
+    debugDrawMutex.lock();
+    requests.push_back(DebugDrawRequest{objSphere, wireframeRasterState,
+                                        glm::translate(center) * glm::scale(glm::vec3(radius)), time, color});
+    debugDrawMutex.unlock();
+}
+
+void Debug::flush(glm::mat4 vp, double deltaT)
+{
+    renderDevice->setShaderPipeline(pipeline);
+
+    debugDrawMutex.lock();
+    for (auto it = requests.begin(); it != requests.end();)
+    {
+        renderDevice->setVertexArray(it->obj.va);
+        renderDevice->setIndexBuffer(it->obj.ib);
+        mvpBindingPoint->bind(mvpBuffer);
+
+        auto& mvp = *(glm::mat4*)mvpBuffer->map();
+        mvp = vp * it->modelMatrix;
+        mvpBuffer->unmap();
+
+        colorBindingPoint->setConstant(it->color);
+
+        renderDevice->setRasterState(it->rasterState);
+        renderDevice->drawTrianglesIndexed(0, it->obj.numIndices);
+
+        it->timeLeft -= deltaT;
+
+        auto current = it++;
+        if (current->timeLeft <= 0)
+        {
+            requests.erase(current);
+        }
+    }
+    debugDrawMutex.unlock();
+    renderDevice->setRasterState(nullptr);
```

Comment:
Unnecessary, you can remove this if you want. Other rendering code should be responsible for setting its rasterizer state, so there's no need to have an extra call here.